numpy/core/src/multiarray/descriptor.c

```diff
@@ -1450,6 +1461,23 @@ _convert_from_type(PyObject *obj) {
         }
         Py_DECREF(ret);
 
+        /*
+         * Handle scalars associated with NEP-42 dtypes
+         */
+
+        if (PyObject_HasAttrString((PyObject *) typ,
+                                   "__associated_array_dtype__")) {
+            PyObject * array_dtype = PyObject_GetAttrString(
+                (PyObject *)typ, "__associated_array_dtype__");
+
+            if (array_dtype != NULL) {
+                ret = ((PyArray_DTypeMeta *)array_dtype)->singleton;
+                Py_INCREF(ret);
+                Py_DECREF(array_dtype);
+                return ret;
+            }
+        }
+
```

Comment:
This section is needed because some of the functions in `np.char` pass the scalar type rather than a dtype instance to `_vec_string`. For example, `np.char.multiply`: https://github.com/numpy/numpy/blob/main/numpy/core/defchararray.py#L373-L374

Reply:
Sorry for only looking at this closer now.  I have to admit, I am not quite convinced that adding more logic here is the best path.  I would be happy to allow e.g. `type(np.dtype("S"))(8)` to succeed (allow instantiating these legacy dtypes with a length from their type).

We pass `(type, size)` to `_vec_string`, but it seems to me that these are few cases where the code could be easily changed to `type(dtype)(size)` instead.